### PR TITLE
feat: add shorthand `{}` for sub-routers

### LINKED ISSUE #3630: feat: add shorthand `{}` for sub-routers
### Describe the feature you'd like to request

I don't wanna write `router({ })` for my sub-routers

### Describe the solution you'd like to see

These should be equivalent

```ts
const appRouter = router({
  foo: router({
    bar: publicProcedure.query(() => '...'),
  })
})
```

```ts
const appRouter = router({
  foo: {
    bar: publicProcedure.query(() => '...'),
  }
})
```

### Describe alternate solutions

Not doing it

### Additional information

If it creates tech debt, don't do it.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

<sub>[TRP-49](https://linear.app/trpc/issue/TRP-49/feat-add-shorthand-for-sub-routers)</sub>

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/3630">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/3630/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by Nsttt**
Closes #3630 

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [x] If necessary, I have added documentation related to the changes made.
- [x] I have added or updated the tests related to the changes made.

### CODE REVIEW

**Comment by Nsttt** [comment]
I'm a bit stuck with the current implementation.

In `router.test.ts` foo is infered as `never` which makes me think somewhere in the `router.ts` file I need to recursively call `CreateRouterInner` in order to get a proper type signature on nested routers.

Current signature.
```ts
const router: CreateRouterInner<$Config, {
    foo: {
        bar: BuildProcedure<"query", {
            _config: $Config;
            ... 5 more ...;
            _meta: object;
        }, string>;
    };
}>
```

What I think it should be.
```ts
const router: CreateRouterInner<$Config, {
    foo: CreateRouterInner<$Config, {
        bar: BuildProcedure<"query", {
            _config: $Config;
            ... 5 more ...;
            _meta: object;
        }, string>;
    }>
}>
```

**Comment by ixahmedxi** [comment]
Hey any updates on this @Nsttt ? I can jump on and help out :)

**Comment by Nsttt** [comment]
> Hey any updates on this @Nsttt ? I can jump on and help out :)

Yeah I'll appreciate the help I'll hit you up on discord ❤️

**Comment by kamilogorek** [comment]
You can apply the patch below automatically from the root of your local repo (make sure it clean first) with:

```terminal
curl https://gist.githubusercontent.com/kamilogorek/87497d8eea109f6a87443392fe34e586/raw/1ad6346cfb54420c05e6325978f21e3debdebbb1/nested.patch | git apply -
```

You are welcome :)

```patch
diff --git ./packages/server/src/core/router.ts ./packages/server/src/core/router.ts
index 4c7f3112..99316b4c 100644
--- ./packages/server/src/core/router.ts
+++ ./packages/server/src/core/router.ts
@@ -97,9 +97,9 @@ type DecorateProcedure<TProcedure extends AnyProcedure> = (
  * @internal
  */
 type DecoratedProcedureRecord<TProcedures extends ProcedureRouterRecord> = {
-  [TKey in keyof TProcedures]: TProcedures[TKey] extends
-    | AnyRouter
-    | ProcedureRouterRecord
+  [TKey in keyof TProcedures]: TProcedures[TKey] extends ProcedureRouterRecord
+    ? DecoratedProcedureRecord<TProcedures[TKey]>
+    : TProcedures[TKey] extends AnyRouter
     ? DecoratedProcedureRecord<TProcedures[TKey]['_def']['record']>
     : TProcedures[TKey] extends AnyProcedure
     ? DecorateProcedure<TProcedures[TKey]>
@@ -150,10 +150,7 @@ function isRouter(
 function isNestedRouter(
   procedureOrRouter: AnyProcedure | AnyRouter | ProcedureRouterRecord,
 ): procedureOrRouter is ProcedureRouterRecord {
-  return (
-    !('router' in procedureOrRouter._def) &&
-    !procedureOrRouter.hasOwnProperty('_procedure')
-  );
+  return !('_def' in procedureOrRouter);
 }
 
 const emptyRouter = {
@@ -215,13 +212,13 @@ export function createRouterFactory<TConfig extends AnyRootConfig>(
     for (const [key, procedureOrRouter] of Object.entries(procedures ?? {})) {
       const value = procedures[key] ?? {};
 
-      if (isRouter(value)) {
-        newProcedures[key] = procedureOrRouter;
+      if (isNestedRouter(value)) {
+        newProcedures[key] = createRouterInner(value);
         continue;
       }
 
-      if (isNestedRouter(value)) {
-        newProcedures[key] = createRouterInner(value);
+      if (isRouter(value)) {
+        newProcedures[key] = procedureOrRouter;
         continue;
       }
 
@@ -233,13 +230,13 @@ export function createRouterFactory<TConfig extends AnyRootConfig>(
       for (const [key, procedureOrRouter] of Object.entries(procedures ?? {})) {
         const newPath = `${path}${key}`;
 
-        if (isRouter(procedureOrRouter)) {
-          recursiveGetPaths(procedureOrRouter._def.procedures, `${newPath}.`);
+        if (isNestedRouter(procedureOrRouter)) {
+          recursiveGetPaths(procedureOrRouter, `${newPath}.`);
           continue;
         }
 
-        if (isNestedRouter(procedureOrRouter)) {
-          recursiveGetPaths(procedureOrRouter, `${newPath}.`);
+        if (isRouter(procedureOrRouter)) {
+          recursiveGetPaths(procedureOrRouter._def.procedures, `${newPath}.`);
           continue;
         }
 ```

**Comment by Nsttt** [comment]
Oh wow it was that simple. GREAT.

I'll try it out later when I arrive home. Many thanks ❤️

**Comment by Nsttt** [comment]
I think we may need some more tests to check this works properly.
Would be down to implement any you find necessary. 

Hit me up with ideas!

**Comment by ixahmedxi** [comment]
@Nsttt i added two more tests but not sure how effective they are

File: packages/tests/server/router.test.ts [hunk 1]
```diff
@@ -86,4 +87,75 @@ describe('router', () => {
     const result = await caller.foo.bar.foo.bar.foo.bar();
     expect(result).toBe('Hello I am recursive');
   });
+
+  test('combined router object should be accessible', async () => {
```
Comment (kamilogorek) [inline packages/tests/server/router.test.ts:91]: The name of the test implies that the setup would be something like

```js
const bar = t.router({
  foo: t.procedure.query(() => 'Hello I am recursive'),
});

const router = t.router({
  foo: {
    bar,
  },
});
```

In its current shape, it's no different than `multiple nested levels of subrouter should be accessible` one.
Although I believe that even the case I described above is already covered by `multiple nested levels of subrouter with different constructors should be accessible`.

**Review by KATT — APPROVED** [review]
let's goooo

**Review by KATT — DISMISSED** [review]

**Review by KATT — APPROVED** [review]

**Pull request merged**